<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="AGC, AY" />


<title>WGCNA</title>

<script src="site_libs/header-attrs-2.9.7/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RAISE_GENIC</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/KJPMolgenLab/RAISE_genic">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">WGCNA</h1>
<h4 class="author">AGC, AY</h4>
<h4 class="date">22 3 2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-09-14
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>RAISE_GENIC/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210706code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210706)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210706code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210706)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomKJPMolgenLabRAISEgenictree7330ccdd33f0eab5e85237bf0f8a3ca462a52fe0targetblank7330ccda"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/KJPMolgenLab/RAISE_genic/tree/7330ccdd33f0eab5e85237bf0f8a3ca462a52fe0" target="_blank">7330ccd</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomKJPMolgenLabRAISEgenictree7330ccdd33f0eab5e85237bf0f8a3ca462a52fe0targetblank7330ccda" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/KJPMolgenLab/RAISE_genic/tree/7330ccdd33f0eab5e85237bf0f8a3ca462a52fe0" target="_blank">7330ccd</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/CountmatrixA.RData
    Ignored:    data/CountmatrixB.RData
    Ignored:    data/Drugdata/
    Ignored:    data/Epi25EE-DEUPUM_DATA_2021-01-21_0748.csv
    Ignored:    data/Epi25EE-DEUPUM_R_2021-01-21_0748.r
    Ignored:    data/Epi25EE_DataDictionary_2020-06-25.csv
    Ignored:    data/Epi25Focal-DEUPUM_DATA_2021-01-21_0746.csv
    Ignored:    data/Epi25Focal-DEUPUM_R_2021-01-21_0746.r
    Ignored:    data/Epi25Focal_DataDictionary_2020-06-25.csv
    Ignored:    data/Epi25GGEOrMixedGener-DEUPUM_DATA_2021-01-21_0744.csv
    Ignored:    data/Epi25GGEOrMixedGener-DEUPUM_R_2021-01-21_0744.r
    Ignored:    data/Epi25GGEOrMixedGeneralisedAndF_DataDictionary_2020-06-25.csv
    Ignored:    data/RAISEGENIC_DATA_2021-01-23_1705.csv
    Ignored:    data/RAISEGENIC_R_2021-01-23_1705.r
    Ignored:    data/VCF_Files/

Untracked files:
    Untracked:  output/GOresWGCNA.xlsx
    Untracked:  output/GOres_Carbamazebine.xlsx
    Untracked:  output/GOres_Carbamazepine.xlsx
    Untracked:  output/GOres_Lacosamide.xlsx
    Untracked:  output/GOres_Levetiracetam.xlsx
    Untracked:  output/GOres_Valporate.xlsx
    Untracked:  output/GOres_Valproate.xlsx
    Untracked:  output/ResTabs_KO.RData
    Untracked:  output/WGCNA_adj_TOM.RData
    Untracked:  output/dds_matrix.RData

Unstaged changes:
    Modified:   setup_start.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<pre class="r"><code>home = getwd()</code></pre>
<div id="wgcna" class="section level1">
<h1>WGCNA</h1>
<div id="identify-gene-modules" class="section level2">
<h2>Identify Gene Modules</h2>
<pre class="r"><code>#calclute the adjacency matrix
allowWGCNAThreads()</code></pre>
<pre><code>Allowing multi-threading with up to 16 threads.</code></pre>
<pre class="r"><code>vsd &lt;- vst(ddsMat, blind=FALSE)
vsd &lt;- assay(vsd)
vsd &lt;- limma::removeBatchEffect(vsd, ddsMat$Batch)

WGCNA_matrix &lt;- t(vsd) #Need to transform for further calculations


if(reanalyze | !file.exists(paste0(output,&quot;/WGCNA_adj_TOM.RData&quot;))){
  
  s = abs(bicor(WGCNA_matrix)) #biweight mid-correlation
  powers = c(c(1:10), seq(from = 12, to=20, by=2))
  sft = pickSoftThreshold(WGCNA_matrix, powerVector = powers, verbose = 5)
  
  par(mfrow = c(1,2))
  cex1 = 0.9;
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab=&quot;Soft Threshold (power)&quot;,ylab=&quot;Scale Free Topology Model Fit, signed R^2&quot;,
       type=&quot;n&quot;, main = paste(&quot;Scale independence&quot;));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col=&quot;red&quot;);
  abline(h=0.90,col=&quot;red&quot;)
  
  plot(sft$fitIndices[,1], sft$fitIndices[,5],xlab=&quot;Soft Threshold (power)&quot;,ylab=&quot;Mean Connectivity&quot;, 
       type=&quot;n&quot;,main = paste(&quot;Mean connectivity&quot;))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col=&quot;red&quot;)
  
  softPower = 8; # The point where the curve flattens
  
  adj= adjacency(WGCNA_matrix,type = &quot;unsigned&quot;, power = softPower);
  
  #Converting adjacency matrix into To so that the noise could be reduced
  TOM=TOMsimilarityFromExpr(WGCNA_matrix,networkType = &quot;unsigned&quot;, 
                            TOMType = &quot;unsigned&quot;, power = softPower);
  save(list = c(&quot;adj&quot;, &quot;TOM&quot;), file=paste0(output,&quot;/WGCNA_adj_TOM.RData&quot;))
  
} else {
  load(paste0(output,&quot;/WGCNA_adj_TOM.RData&quot;))
}</code></pre>
<pre><code>pickSoftThreshold: will use block size 2862.
 pickSoftThreshold: calculating connectivity for given powers...
   ..working on genes 1 through 2862 of 15629
   ..working on genes 2863 through 5724 of 15629
   ..working on genes 5725 through 8586 of 15629
   ..working on genes 8587 through 11448 of 15629
   ..working on genes 11449 through 14310 of 15629
   ..working on genes 14311 through 15629 of 15629
   Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.
1      1    0.139  0.578          0.894 4250.00  4140.000   6930
2      2    0.397 -0.618          0.891 1750.00  1560.000   4150
3      3    0.719 -1.020          0.926  883.00   701.000   2820
4      4    0.787 -1.220          0.928  503.00   350.000   2060
5      5    0.812 -1.350          0.934  310.00   187.000   1570
6      6    0.831 -1.420          0.944  203.00   106.000   1230
7      7    0.846 -1.470          0.955  139.00    62.300    994
8      8    0.857 -1.500          0.963   98.50    38.100    815
9      9    0.857 -1.540          0.964   71.80    23.800    678
10    10    0.856 -1.570          0.964   53.50    15.200    571
11    12    0.858 -1.610          0.968   31.50     6.570    417
12    14    0.878 -1.610          0.982   19.60     3.000    314
13    16    0.886 -1.630          0.985   12.80     1.450    242
14    18    0.896 -1.630          0.989    8.62     0.724    191
15    20    0.902 -1.640          0.991    6.00     0.376    153</code></pre>
<p><img src="figure/04_WGCNA.Rmd/adjacency-1.png" width="4500" style="display: block; margin: auto;" /></p>
<pre><code>TOM calculation: adjacency..
..will not use multithreading.
 Fraction of slow calculations: 0.000000
..connectivity..
..matrix multiplication (system BLAS)..
..normalization..
..done.</code></pre>
<pre class="r"><code>SubGeneNames&lt;-colnames(WGCNA_matrix)

colnames(TOM) =rownames(TOM) =SubGeneNames
dissTOM=1-TOM
diag(dissTOM) = 0

#hierarchical clustering
geneTree = flashClust(as.dist(dissTOM),method=&quot;average&quot;);

#plot the resulting clustering tree (dendrogram)
# plot(geneTree, xlab=&quot;&quot;, sub=&quot;&quot;,cex=0.3, main=&quot;Module clustering prio to merging&quot;);

#Set the minimum module size
minModuleSize = 50;

#Module identification using dynamic tree cut
dynamicMods = cutreeDynamic(dendro = geneTree, 
                            distM = dissTOM, 
                            cutHeight = 0.998,
                            minClusterSize = minModuleSize,
                            deepSplit=2, 
                            pamRespectsDendro = FALSE)</code></pre>
<pre><code> ..done.</code></pre>
<pre class="r"><code>#the following command gives the module labels and the size of each module. Lable 0 is reserved for unassigned genes

#Plot the module assignment under the dendrogram; note: The grey color is reserved for unassigned genes
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)</code></pre>
<pre><code>dynamicColors
         black           blue          brown           cyan      darkgreen 
           496           2594            795            287            175 
      darkgrey    darkmagenta darkolivegreen     darkorange        darkred 
           170            118            118            165            205 
 darkturquoise          green    greenyellow           grey         grey60 
           172            521            401            213            246 
     lightcyan     lightgreen    lightyellow        magenta   midnightblue 
           247            228            225            467            271 
        orange     orangered4  paleturquoise           pink          plum1 
           166             63            130            474             70 
        purple            red      royalblue    saddlebrown         salmon 
           417            511            209            136            378 
       sienna3        skyblue       skyblue3      steelblue            tan 
           105            154             89            130            397 
     turquoise         violet          white         yellow    yellowgreen 
          2989            129            158            718             92 </code></pre>
<pre class="r"><code>plotDendroAndColors(geneTree, dynamicColors, &quot;Dynamic Tree Cut&quot;, 
                    dendroLabels = FALSE, hang = 0.03, 
                    addGuide = TRUE, 
                    guideHang = 0.05, 
                    main = &quot;Gene dendrogram and module colors&quot;)</code></pre>
<p><img src="figure/04_WGCNA.Rmd/adjacency-2.png" width="4500" style="display: block; margin: auto;" /></p>
<pre class="r"><code>diag(dissTOM) = NA;

#Visualize the Tom plot. Raise the dissimilarity matrix to the power of 4 to bring out the module structure
TOMplot(dissTOM^4, geneTree, as.character(dynamicColors), main=&quot;weighted distance of Topological overlap Matrix&quot;)</code></pre>
<p><img src="figure/04_WGCNA.Rmd/adjacency-3.png" width="4500" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="module-eigengenes" class="section level1">
<h1>Module eigengenes</h1>
<div id="identify-and-merge-correlated-modules" class="section level2">
<h2>Identify and Merge correlated modules</h2>
<pre class="r"><code>#calculate eigengenes
MEList = moduleEigengenes(WGCNA_matrix, colors = dynamicColors)
MEs = MEList$eigengenes 
plotEigengeneNetworks(MEs, &quot;Eigengenes_before_merging&quot;, marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2))</code></pre>
<p><img src="figure/04_WGCNA.Rmd/MEs-1.png" width="7500" style="display: block; margin: auto;" /></p>
<pre class="r"><code>MEList_new = mergeCloseModules(WGCNA_matrix, colors = dynamicColors, MEs = MEs, cutHeight = 0.2)</code></pre>
<pre><code> mergeCloseModules: Merging modules whose distance is less than 0.2
   Calculating new MEs...</code></pre>
<pre class="r"><code>plotEigengeneNetworks(MEList_new$newMEs, &quot;Eigengenes_after_merging&quot;, 
                      marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2))</code></pre>
<p><img src="figure/04_WGCNA.Rmd/MEs-2.png" width="7500" style="display: block; margin: auto;" /></p>
<pre class="r"><code>MEs = MEList_new$newMEs

colData(ddsMat) = colData(ddsMat) %&gt;% cbind(MEs[colnames(ddsMat),])

colors_new = MEList_new$colors

mcols(ddsMat) = cbind(mcols(ddsMat), data.frame(cluster=colors_new))

save(ddsMat, file=paste0(output,&quot;/dds_matrix.RData&quot;))</code></pre>
</div>
<div id="heatmaps-mes" class="section level2">
<h2>Heatmaps MEs</h2>
<pre class="r"><code>#colors for plotting heatmap
#colors for plotting heatmap
colors &lt;- colorRampPalette(brewer.pal(9, &quot;Spectral&quot;))(255)

cellcol = Dark8[1:nlevels(SampleInfo$CellLine)]
names(cellcol) = levels(SampleInfo$CellLine)

drugcol = Dark8[c(1:nlevels(SampleInfo$Drug)+nlevels(SampleInfo$CellLine))]
names(drugcol) = levels(SampleInfo$Drug)


batchcol = Dark8[c(1:nlevels(SampleInfo$Batch)+nlevels(SampleInfo$CellLine)+nlevels(SampleInfo$Drug))]
names(batchcol) = levels(SampleInfo$Batch)

conzcol = c(&quot;orange&quot;, &quot;darkred&quot;)

ann_colors = list(
  CellLine = cellcol, 
  Drug = drugcol,
  Batch = batchcol,
  Conz=conzcol)



idx=order(SampleInfo$CellLine, SampleInfo$Drug, SampleInfo$Conz,SampleInfo$Batch)
WGCNA_matrix_sorted=WGCNA_matrix[idx, order(colors_new)]
rownames(WGCNA_matrix_sorted) = SampleInfo$label_rep[idx]

collabels = SampleInfo[idx,c(&quot;CellLine&quot;,&quot;Drug&quot;,&quot;Conz&quot;,&quot;Batch&quot;)] %&gt;%  
  mutate_if(is.factor, as.character) %&gt;% as.data.frame()

rownames(collabels)=SampleInfo$label_rep[idx]

genlabels = data.frame(cluster = as.character(colors_new)[order(colors_new)])
rownames(genlabels) = colnames(WGCNA_matrix_sorted)

pheatmap(t(WGCNA_matrix_sorted),
         cluster_rows= F,
         cluster_cols = F,
         border_color = NA, 
         clustering_method = &quot;ward.D2&quot;,
         annotation_row = genlabels,
         annotation_col = collabels,
         annotation_colors = ann_colors,
         col = colors, 
         scale = &quot;row&quot;,
         main = &quot;Distances  normalized log2 counts&quot;)</code></pre>
<p><img src="figure/04_WGCNA.Rmd/heatmap-1.png" width="7500" style="display: block; margin: auto;" /></p>
<pre class="r"><code>MElabels = data.frame(cluster = gsub(&quot;ME&quot;, &quot;&quot;,colnames(MEs)))
rownames(MElabels) = colnames(MEs)

clustcol = gplots::col2hex(unique(as.character(MElabels$cluster)))
names(clustcol) = as.character(MElabels$cluster)


rownames(MEs) = SampleInfo[rownames(MEs),&quot;label_rep&quot;]

pheatmap(t(MEs[idx,]),
         border_color = NA, 
         annotation_row = MElabels,
         cluster_cols = F,
         annotation_col = collabels,
         clustering_method = &quot;ward.D2&quot;,
         annotation_colors = ann_colors,
         scale=&quot;none&quot;,
         col = colors, 
         main = &quot;Distances  normalized log2 counts&quot;)</code></pre>
<p><img src="figure/04_WGCNA.Rmd/heatmap-2.png" width="7500" style="display: block; margin: auto;" /></p>
<pre class="r"><code>gene_univers = rownames(ddsMat)

Genes_of_interset = split(rownames(ddsMat), mcols(ddsMat)$cluster)

gostres = getGOresults(Genes_of_interset, gene_univers)

toptab = gostres$result

write.xlsx2(toptab, file = paste0(output,&quot;/GOresWGCNA.xlsx&quot;), sheetName = &quot;GO_enrichment&quot;)

for (module in names(Genes_of_interset)){
  idx = toptab$query==module &amp; grepl(&quot;GO&quot;, toptab$source)
  if(!any(idx)){
    p = ggplot() + annotate(&quot;text&quot;, x = 4, y = 25, size=4, 
                            label = &quot;no significant GO term&quot;) + 
      ggtitle(module)+theme_void()+
      theme(plot.title = element_text(hjust = 0.5))
  } else {
    p=GOplot(toptab[idx, ], 10, Title =module)
  }
  print(p)
}</code></pre>
<p><img src="figure/04_WGCNA.Rmd/goresplots-1.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-2.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-3.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-4.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-5.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-6.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-7.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-8.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-9.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-10.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-11.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-12.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-13.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-14.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-15.png" width="4500" style="display: block; margin: auto;" /><img src="figure/04_WGCNA.Rmd/goresplots-16.png" width="4500" style="display: block; margin: auto;" /></p>
<pre class="r"><code>SampleInfo=as.data.frame(colData(ddsMat))
MEnames = grep(&quot;ME&quot;, colnames(SampleInfo), value = T)
randomeffect = c(&quot;Batch&quot;, &quot;CellLine&quot;)

target = &quot;Conz&quot;

for (drug in unique(SampleInfo$Drug)){
  res=data.frame()
  for (mod in MEnames){
    re = paste0(&quot;(1|&quot;, randomeffect, &quot;)&quot;, collapse=&quot;+&quot;)
    y = SampleInfo[,mod]
    model = paste0(&quot;y~&quot;,re,&quot;+&quot;, target)
    data = SampleInfo[SampleInfo$Drug==drug,c(randomeffect, target)]
    y=y[SampleInfo$Drug==drug]
    resultlme = lmer(model, data)
    stats = summary(resultlme)$coeff[2,]
    res = rbind(res, stats)
    colnames(res) = colnames(summary(resultlme)$coeff)
  }
  rownames(res) = MEnames
  res$p.value = 2*pt(res$`t value`, sum(SampleInfo$Drug==drug), lower=FALSE)
  res$p.value[res$p.value&gt;1]=1
  assign(paste0(&quot;WGCNA_lmer_&quot;, drug), res)
}</code></pre>
<pre><code>boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular
boundary (singular) fit: see ?isSingular</code></pre>
<pre class="r"><code># 
# boxplot(SampleInfo[SampleInfo$Drug==&quot;Valproate&quot;, &quot;MEdarkturquoise&quot;] ~ as.factor(SampleInfo[SampleInfo$Drug==&quot;Valporate&quot;, &quot;Conz&quot;]), 
#          ylab=&quot;MEdarkturquoise&quot;, xlab=&quot;mM&quot;, col=&quot;darkturquoise&quot;)
# 
# boxplot(SampleInfo[SampleInfo$Drug==&quot;Levetiracetam&quot;, &quot;MEcyan&quot;] ~ as.factor(SampleInfo[SampleInfo$Drug==&quot;Valporate&quot;, &quot;Conz&quot;]), 
#          ylab=&quot;MEcyan&quot;, xlab=&quot;mM&quot;, col=&quot;cyan&quot;)

for (i in apropos(&quot;WGCNA_lmer_&quot;)){
  i
  
}

save(list=apropos(&quot;WGCNA_lmer_&quot;), file=paste0(output,&quot;/WGCNA_lmer_res.RData&quot;))</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.4 (2021-02-15)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252   
[3] LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   
[5] LC_TIME=German_Germany.1252    

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] lme4_1.1-26                 Matrix_1.3-2               
 [3] xlsx_0.6.5                  gprofiler2_0.2.0           
 [5] flashClust_1.01-2           WGCNA_1.70-3               
 [7] fastcluster_1.1.25          dynamicTreeCut_1.63-1      
 [9] pheatmap_1.0.12             knitr_1.33                 
[11] limma_3.46.0                DESeq2_1.30.1              
[13] SummarizedExperiment_1.20.0 Biobase_2.50.0             
[15] MatrixGenerics_1.2.1        matrixStats_0.58.0         
[17] GenomicRanges_1.42.0        GenomeInfoDb_1.26.4        
[19] IRanges_2.24.1              S4Vectors_0.28.1           
[21] BiocGenerics_0.36.0         RColorBrewer_1.1-2         
[23] compareGroups_4.5.1         forcats_0.5.1              
[25] stringr_1.4.0               dplyr_1.0.5                
[27] purrr_0.3.4                 readr_1.4.0                
[29] tidyr_1.1.3                 tibble_3.1.0               
[31] ggplot2_3.3.5               tidyverse_1.3.0            
[33] kableExtra_1.3.4            workflowr_1.6.2            

loaded via a namespace (and not attached):
  [1] utf8_1.1.4             tidyselect_1.1.0       RSQLite_2.2.4         
  [4] AnnotationDbi_1.52.0   htmlwidgets_1.5.3      grid_4.0.4            
  [7] BiocParallel_1.24.1    munsell_0.5.0          codetools_0.2-18      
 [10] preprocessCore_1.52.1  chron_2.3-56           statmod_1.4.35        
 [13] withr_2.4.1            colorspace_2.0-0       highr_0.9             
 [16] uuid_0.1-4             rstudioapi_0.13        officer_0.3.17        
 [19] rJava_0.9-13           labeling_0.4.2         git2r_0.28.0          
 [22] GenomeInfoDbData_1.2.4 bit64_4.0.5            farver_2.1.0          
 [25] rprojroot_2.0.2        vctrs_0.3.6            generics_0.1.0        
 [28] xfun_0.24              R6_2.5.0               doParallel_1.0.16     
 [31] locfit_1.5-9.4         bitops_1.0-6           cachem_1.0.4          
 [34] DelayedArray_0.16.2    assertthat_0.2.1       promises_1.2.0.1      
 [37] scales_1.1.1           nnet_7.3-15            gtable_0.3.0          
 [40] rlang_0.4.10           genefilter_1.72.1      systemfonts_1.0.1     
 [43] splines_4.0.4          lazyeval_0.2.2         impute_1.64.0         
 [46] broom_0.7.5            checkmate_2.0.0        yaml_2.2.1            
 [49] modelr_0.1.8           backports_1.2.1        httpuv_1.5.5          
 [52] HardyWeinberg_1.7.1    Hmisc_4.5-0            tools_4.0.4           
 [55] ellipsis_0.3.1         gplots_3.1.1           jquerylib_0.1.3       
 [58] Rsolnp_1.16            Rcpp_1.0.6             base64enc_0.1-3       
 [61] zlibbioc_1.36.0        RCurl_1.98-1.2         rpart_4.1-15          
 [64] haven_2.3.1            cluster_2.1.1          fs_1.5.0              
 [67] magrittr_2.0.1         data.table_1.14.0      flextable_0.6.4       
 [70] reprex_1.0.0           truncnorm_1.0-8        hms_1.0.0             
 [73] xlsxjars_0.6.1         evaluate_0.14          xtable_1.8-4          
 [76] XML_3.99-0.5           jpeg_0.1-8.1           readxl_1.3.1          
 [79] gridExtra_2.3          compiler_4.0.4         mice_3.13.0           
 [82] KernSmooth_2.23-18     writexl_1.3.1          crayon_1.4.1          
 [85] minqa_1.2.4            htmltools_0.5.1.1      later_1.1.0.1         
 [88] Formula_1.2-4          geneplotter_1.68.0     lubridate_1.7.10      
 [91] DBI_1.1.1              dbplyr_2.1.0           MASS_7.3-53.1         
 [94] boot_1.3-27            cli_3.0.1              pkgconfig_2.0.3       
 [97] foreign_0.8-81         plotly_4.9.3           xml2_1.3.2            
[100] foreach_1.5.1          svglite_2.0.0          annotate_1.68.0       
[103] bslib_0.2.4            webshot_0.5.2          XVector_0.30.0        
[106] rvest_1.0.0            digest_0.6.27          rmarkdown_2.9.7       
[109] cellranger_1.1.0       htmlTable_2.1.0        gdtools_0.2.3         
[112] gtools_3.8.2           nloptr_1.2.2.2         lifecycle_1.0.0       
[115] nlme_3.1-152           jsonlite_1.7.2         viridisLite_0.3.0     
[118] fansi_0.4.2            pillar_1.5.1           lattice_0.20-41       
[121] fastmap_1.1.0          httr_1.4.2             survival_3.2-9        
[124] GO.db_3.12.1           glue_1.4.2             zip_2.1.1             
[127] png_0.1-7              iterators_1.0.13       bit_4.0.4             
[130] stringi_1.7.3          sass_0.3.1             blob_1.2.1            
[133] latticeExtra_0.6-29    caTools_1.18.2         memoise_2.0.0         </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
