---
title: "Preprocessing_Countdata"
author: "AGC"
date: "27 8 2021"
output: html_document
---


```{r}
home = getwd()
```


```{r setup, include=FALSE, echo=FALSE}
secret="C:/Users/andreas_chiocchetti/OneDrive/personal/fuchs_credentials.R"


output= paste0(home, "/output/")
rankconz = T

source(paste0(home,"/code/00_installer.R")) # installs pacakges if needed
source(paste0(home,"/code/custom_functions.R")) #also defines colors

library(knitr)
library(RCurl) 
library(DESeq2)
library(RColorBrewer)
library(glmpca)
library(ggplot2)
library(tidyverse)
library(pheatmap)
library(dendextend)
library(gprofiler2)
library(readxl)

opts_chunk$set(echo = TRUE, 
               fig.height = 7, 
               fig.width = 9, 
               message = TRUE,
               warning = TRUE,
               fig.align='center',
               dpi=500
)

options(stringsAsFactors = F)


set.seed(157434538)


```

## Preprocessing BATCH A 


```{r getdata BATCH_A}

#Load Count Matrix
if (file.exists(secret)){
  source(secret)
} else {
  UID    = rstudioapi::askForPassword("fuchs user")
  PWD    = rstudioapi::askForPassword("fuchs password")
}

filetarget= paste0(home,"/data/CountmatrixA.RData")
if(!file.exists(filetarget)){
  url="sftp://fuchs.hhlr-gu.de//scratch/fuchs/agmisc/chiocchetti/RNASeq_Data/RAISEGenic/Rawfiles/BatchA/Countmatrix/Countmatrix.RData"
  bin = getBinaryURL(url, userpw=paste0(UID,":",PWD))
  writeBin(bin, filetarget)
  load(filetarget)
} else {
  load(filetarget)
}

Ntot= nrow(Countdata)


#merge non unique annotations
if(length(unique(rownames(Countdata))) != Ntot){
  Countdata = Countdata %>% group_by(row.names(Countdata)) %>% summarise_each(sum)
  Ntot= nrow(Countdata)
}


hgnc=gconvert(query=as.numeric(rownames(Countdata)), 
              organism = "hsapiens", 
              numeric_ns = "ENTREZGENE_ACC",
              target = "HGNC")

Ids = hgnc %>%  dplyr::select(name, input, description) %>% group_by(input) %>% 
  summarise(name=paste(name, sep="; ", collapse = ";"), description = dplyr::first(description))

rowdescription = data.frame(entrez_gene = Ids$input, 
                            hgnc=Ids$name, 
                            description=Ids$description)


if(all(row.names(Countdata) %in% rowdescription$entrez_gene)){
  
  rowdescription = rowdescription[match(row.names(Countdata), rowdescription$entrez_gene),]
  rownames(rowdescription)=row.names(Countdata)
}


# load and parse sample information 
SampleInfo=read_xlsx(paste0(home,"/data/Drugdata/20210630_2021-090-LEX/P2021-090-LEX_Probenübersicht.xlsx"))
SampleInfo=as.data.frame(SampleInfo)

# set factors and relevel
SampleInfo$CellLine = as.factor(SampleInfo$CellLine)

SampleInfo$Drug = factor(SampleInfo$Drug, levels=c("CBZ","LCA","LTA","VPA"), 
                         labels=c("Carbamazepine", "Lacosamide", "Levetiracetam", "Valproate"))

rownames(SampleInfo) = paste0(SampleInfo$ID,"a")

# align datasets
checkfiles = all(rownames(SampleInfo) %in% colnames(Countdata))
IDs=intersect(rownames(SampleInfo), colnames(Countdata))
Countdata = Countdata[,IDs]
SampleInfo = SampleInfo[IDs, ]

SampleInfo$reads_per_sample = colSums(Countdata)

display_tab(head(Countdata))
display_tab(SampleInfo)

Countdata.A = Countdata
SampleInfo.A = SampleInfo
SampleInfo.A$Batch = "A"

```
Total number of samples overlapping between Counts and SampleInfo: `r nrow(SampleInfo)`


## Preprocessing BATCH B 

```{r getdata BATCH_B}

#Load Count Matrix
if (file.exists(secret)){
  source(secret)
} else {
  UID    = rstudioapi::askForPassword("fuchs user")
  PWD    = rstudioapi::askForPassword("fuchs password")
}

filetarget= paste0(home,"/data/CountmatrixB.RData")
if(!file.exists(filetarget)){
  url="sftp://fuchs.hhlr-gu.de//scratch/fuchs/agmisc/chiocchetti/RNASeq_Data/RAISEGenic/Rawfiles/BatchB/Countmatrix/Countmatrix.RData"
  bin = getBinaryURL(url, userpw=paste0(UID,":",PWD))
  writeBin(bin, filetarget)
  load(filetarget)
} else {
  load(filetarget)
}

Ntot= nrow(Countdata)


#merge non unique annotations
if(length(unique(rownames(Countdata))) != Ntot){
  Countdata = Countdata %>% group_by(row.names(Countdata)) %>% summarise_each(sum)
  Ntot= nrow(Countdata)
}

hgnc=gconvert(query=as.numeric(rownames(Countdata)), 
              organism = "hsapiens", 
              numeric_ns = "ENTREZGENE_ACC",
              target = "HGNC")

Ids = hgnc %>%  dplyr::select(name, input, description) %>% group_by(input) %>% 
  summarise(name=paste(name, sep="; ", collapse = ";"), description = dplyr::first(description))

rowdescription = data.frame(entrez_gene = Ids$input, 
                            hgnc=Ids$name, 
                            description=Ids$description)


if(all(row.names(Countdata) %in% rowdescription$entrez_gene)){
  
  rowdescription = rowdescription[match(row.names(Countdata), rowdescription$entrez_gene),]
  rownames(rowdescription)=row.names(Countdata)
}


# load and parse sample information 
SampleInfo=read_xlsx(paste0(home,"/data/Drugdata/20210630_2021-090-LEX/P2021-090-LEX_Probenübersicht.xlsx"))
SampleInfo=as.data.frame(SampleInfo)

# set factors and relevel
SampleInfo$CellLine = as.factor(SampleInfo$CellLine)

SampleInfo$Drug = factor(SampleInfo$Drug, levels=c("CBZ","LCA","LTA","VPA"), 
                         labels=c("Carbamazepine", "Lacosamide", "Levetiracetam", "Valproate"))

rownames(SampleInfo) = paste0(SampleInfo$ID,"b")

# align datasets
checkfiles = all(rownames(SampleInfo) %in% colnames(Countdata))
IDs=intersect(rownames(SampleInfo), colnames(Countdata))
Countdata = Countdata[,IDs]
SampleInfo = SampleInfo[IDs, ]

SampleInfo$reads_per_sample = colSums(Countdata)

display_tab(head(Countdata))
display_tab(SampleInfo)

Countdata.B = Countdata
SampleInfo.B = SampleInfo
SampleInfo.B$Batch = "B"

```

```{r mergebatches}

Countdata = merge(Countdata.A, Countdata.B, by=0, all=TRUE)

SampleInfo = rbind(SampleInfo.A, SampleInfo.B)
Countdata = Countdata %>% column_to_rownames("Row.names")

SampleInfo$label = as.factor(SampleInfo$label)
SampleInfo$reads_per_sample = colSums(Countdata)
SampleInfo$Batch = as.factor(SampleInfo$Batch)
SampleInfo$CellLine = as.factor(SampleInfo$CellLine)
SampleInfo$Drug = as.factor(SampleInfo$Drug)
SampleInfo$label_rep = paste0(SampleInfo$label, "_", SampleInfo$Batch)

SampleInfo = SampleInfo[order(SampleInfo$Drug,SampleInfo$Conz,  SampleInfo$CellLine, SampleInfo$Batch),]
Countdata = Countdata[,rownames(SampleInfo)]

```



## Merged Datasets

```{r plot_counts}

boxplot_counts = function(plotsubset, maintitle, colorcode) {
  par(mar=c(3,3,5,7))
  a =boxplot(log2(plotsubset+1), main = maintitle, 
             col = Dark8[as.factor(SampleInfo[,colorcode])], names=NA,
             ylab = "log2 transformed", xlab="samples")
  legend("topleft", legend = levels(SampleInfo[,colorcode]), bg="white",
         pch = 16, col = Dark8[1:length(unique(SampleInfo[,colorcode]))])
}


barplot_counts = function(DF, maintitle, colorcode) {
  barplot(log2(DF[,"reads_per_sample"]), main = maintitle, 
          col = Dark8[as.factor(DF[,colorcode])], names="",
          ylab = "log2 transformed",xlab="samples")
  legend("bottomleft", legend = levels(DF[,colorcode]), pch = 16, 
         bg="white",
         col = Dark8[1:length(unique(DF[,colorcode]))])
}

par(mfrow=c(1,1))
boxplot_counts(Countdata, "raw counts", "CellLine")
boxplot_counts(Countdata, "raw counts", "Drug")
boxplot_counts(Countdata, "raw counts", "Batch")

barplot_counts(SampleInfo, "total reads", "label")
barplot_counts(SampleInfo, "total reads", "Drug")
barplot_counts(SampleInfo, "total reads", "Batch")


plot(density(log2(rowMeans(Countdata))), main="distribution of gene expression", 
     xlab="mean log2(counts +1)")

# remove genes wich were not detected in at least 50% of the samples 
keeperidx = rowSums(Countdata>1)>nrow(SampleInfo)/2

Countdata_cl = Countdata[keeperidx, ]

rowdescription = rowdescription[row.names(Countdata_cl),]


fullmodel = as.formula("~Batch+CellLine+Drug*Conz")


ddsMat <- DESeqDataSetFromMatrix(countData = Countdata_cl,
                                 colData = SampleInfo,
                                 rowData = rowdescription,
                                 design = fullmodel)

ddsMat = ddsMat[rowSums(counts(ddsMat)) > (2 * ncol(ddsMat)), ]

ddsMat = estimateSizeFactors(ddsMat)
ddsMat = estimateDispersions(ddsMat)

reads  = as.data.frame(counts(ddsMat, normalized=T))

SDs = apply(reads, 1, sd)
keepvar = SDs>0

ddsMat <- ddsMat[keepvar,]

Nfilt = length(ddsMat)
reads  = as.data.frame(counts(ddsMat, normalized=T))

SampleInfo$reads_per_sample_cl= colSums(reads)


```

### before cleaning
* Average reads per samples: `r format(mean(SampleInfo$reads_per_sample), big.mark=" ", scientific=F)`
* Standard deviation reads per samples: `r format(sd(SampleInfo$reads_per_sample), big.mark=" ", scientific=F)`
* Total genes mapped: `r format(Ntot, big.mark=" ")`

### after cleaning

* Average reads per samples: `r format(mean(colSums(reads)), big.mark=" ", scientific=F)`
* Standard deviation reads per samples: `r format(sd(colSums(reads)), big.mark=" ", scientific=F)`
* Genes removed due to low reads: `r format(Ntot-Nfilt, big.mark = " ")` 
* Total genes included after filtering: `r format(Nfilt, big.mark=" ")`


## Clustering 

hierarchical clustering based on the top 2000 genes by variance removing batch effects

```{r hcluster, fig.height=10, fig.width=15}

vsd <- vst(ddsMat, blind=FALSE)
mat <- assay(vsd)
log2_cpm <- limma::removeBatchEffect(mat, vsd$Batch, vsd$CellLine)


varsset=apply(log2_cpm, 1, var)

cpm.sel.trans = t(log2_cpm[order(varsset,decreasing = T)[1:2000],])

rownames(cpm.sel.trans)=SampleInfo$label_rep

distance = dist(cpm.sel.trans)
hc = stats::hclust(distance, method = "ward.D2")

cutN=24
clusters = cutree(hc, k=cutN)
Colors=data.frame(Cellline = Dark8[as.numeric(SampleInfo$CellLine)],
                  Drug = Dark8[as.numeric(SampleInfo$Drug)+nlevels(SampleInfo$CellLine)],
                  Batch = Dark8[as.numeric(SampleInfo$Batch)+nlevels(SampleInfo$CellLine)++nlevels(SampleInfo$Drug)])

par(mar=c(15,3,5,3))
plot(as.dendrogram(hc), main=paste("Similarity by gene expression top 2000"), cex=0.7)
colored_dots(colors = Colors, dend = as.dendrogram(hc))

```


Similarity based on hcluster plot 


```{r clusterandqual, fig.height=15, fig.width=15}
sampleDistMatrix <- as.matrix(distance)

#colors for plotting heatmap
colors <- rev(colorRampPalette(brewer.pal(9, "Spectral"))(255))

cellcol = Dark8[1:nlevels(SampleInfo$CellLine)]
names(cellcol) = levels(SampleInfo$CellLine)

drugcol = Dark8[c(1:nlevels(SampleInfo$Drug)+nlevels(SampleInfo$CellLine))]
names(drugcol) = levels(SampleInfo$Drug)


batchcol = Dark8[c(1:nlevels(SampleInfo$Batch)+nlevels(SampleInfo$CellLine)+nlevels(SampleInfo$Drug))]
names(batchcol) = levels(SampleInfo$Batch)

conzcol = c("orange", "darkred")

ann_colors = list(
  CellLine = cellcol, 
  Drug = drugcol,
  Batch = batchcol,
  Conz=conzcol)


labels = SampleInfo[,c("CellLine","Drug","Conz","Batch")] %>%  
  mutate_if(is.factor, as.character) %>% as.data.frame()
if(rankconz){
  labels$Conz = as.numeric(factor(labels$Conz, levels=c(unique(labels$Conz)), 
                                  labels=c(0,1,2,1,2,1,2)))
}
rownames(labels)=SampleInfo$label_rep
rownames(sampleDistMatrix) = rownames(labels)
colnames(sampleDistMatrix) = rownames(sampleDistMatrix)


pheatmap(sampleDistMatrix,
         clustering_distance_rows = distance,
         clustering_distance_cols = distance,
         clustering_method = "ward.D2",
         scale ="none",
         border_color = NA, 
         annotation_row = labels,
         annotation_col = labels,
         annotation_colors = ann_colors,
         col = colors, 
         main = "Distances  normalized log2 counts")

save(ddsMat, file=paste0(output,"/dds_matrix.RData"))

```


## PCA and MDS

```{r PCA_MDS}
# PCA
gpca <- glmpca(t(cpm.sel.trans), L = 2)
gpca.dat <- gpca$factors
gpca.dat$CellLine <- SampleInfo$CellLine
gpca.dat$Drug <- SampleInfo$Drug
gpca.dat$Batch<- SampleInfo$Batch
gpca.dat$Conz <- SampleInfo$Conz

if(rankconz){
  gpca.dat$Conz = as.numeric(factor(gpca.dat$Conz, levels=c(unique(gpca.dat$Conz)), 
                                     labels=c(0,1,2,1,2,1,2)))
}


rownames(gpca.dat) = SampleInfo$labels
mds = as.data.frame(SampleInfo) %>% cbind(cmdscale(distance))


ggplot(gpca.dat, aes(x = dim1, y = dim2, color = CellLine, shape = Drug, size = Conz)) +
  geom_point()  + ggtitle("PCA with log2 counts")

ggplot(mds, aes(x = `1`, y = `2`, color = CellLine, shape = Drug, size = Conz)) +
  geom_point()  + ggtitle("MDS with log2 counts")


```
